chips/nrf5x/src/ble_advertising_driver.rs

```diff
@@ -148,159 +284,616 @@ impl<'a, B, A> BLE<'a, B, A>
+        self.app
+            .enter(appid, |app, _| {
+                app.advertisement_buf
+                    .as_mut()
+                    .map(|slice| {
+                        slice.as_mut()[PACKET_HDR_PDU] =
+                            BLEAdvertisementType::NonConnectUndirected as u8;
+                        ReturnCode::SUCCESS
+                    })
+                    .unwrap_or_else(|| ReturnCode::ESIZE)
+            })
+            .unwrap_or_else(|err| err.into())
+    }
+
+
     // This function constructs an AD TYPE with type, data, length and offset.
     // It uses the offset to keep track of where to place the next AD TYPE in the buffer in
     // case multiple AD TYPES are provided.
-    // The chip module then sets the actual payload.
-    fn set_adv_data(&self, ad_type: usize) -> ReturnCode {
-        let mut ret = ReturnCode::ESIZE;
-        for cntr in self.app.iter() {
-            cntr.enter(|app, _| {
-                app.app_write.as_ref().map(|slice| {
-                    let len = slice.len();
-                    // Each AD TYP consists of TYPE (1 byte), LENGTH (1 byte) and
-                    // PAYLOAD (0 - 31 bytes)
-                    // This is why we add 2 to start the payload at the correct position.
-                    let i = self.offset.get() + len + 2;
-                    if i <= 31 {
-                        self.kernel_tx.take().map(|data| {
-                            for (out, inp) in data.iter_mut().zip(slice.as_ref()[0..len].iter()) {
-                                *out = *inp;
+    //
+    // But because we borrow the struct mutabily we can't borrow it immutably at the same time
+    // First we copy the data from the allow call to a the TakeCell buffer
+    // And then copy that TakeCell buffer to AppSlice advertisement buffer
+    fn set_advertisement_data(&self, gap_type: BLEGapType, appid: kernel::AppId) -> ReturnCode {
+        // these variables are workaround because we can't access other data members
+        // when we have a mutable borrow!!!
+
+        // keep track of the end data -> update after as the new index afterward
+        let mut end = 0;
+        // index + Buffer length + 2 (1 byte for Length, 1 byte for AD Type)
+        let mut buf_len = 0;
+        // Current index in the buffer
+        let mut index = 0;
+
+        self.app
+            .enter(appid, |app, _| {
+                let status = app.app_write
+                    .as_ref()
+                    .map(|slice| {
+
+                        // get current index
+                        index = app.offset.get();
+                        // get end
+                        end = index + slice.len() + 2;
+                        buf_len = slice.len() + 2;
+
+                        // Copy data from the "WRITE" AppSlice to the TakeCell if there is
+                        // space left
+                        if end <= PACKET_LENGTH {
+                            self.kernel_tx
```

Comment:
I'm pretty confused by this code... What's the point of copying into the kernel buffer at this point? It's just going to be copied back to the `App`'s `advertisement_buf` down below. Moreover, doesn't this mean that if there is currently an advertisement being transmitted (i.e. for a different app) that this code will fail to set the advertisement data for the calling app?

In general, it seems to me like, from an software engineering perspective, several of these functions should actually be methods of the `App` struct and entering the grant should be done by the caller. If nothing else, it would make these methods much more testable independently and easier to read.

Reply:
The point is a workaround for borrowing rules in Rust i.e., not valid to have an immutable borrow to `app_write` and a mutable borrow to `app_advertisement_buf`.
But I can probably move the buffers to a Cell or some other container that provides interior mutability. 
Is Cell ok?

Yes, the code is incredible ugly and hard to read!!

Ohh yes, you are right I figured this was only to be reached once the `busy` flag is set. But this is not the case for example `AllowType::BLEGap` call.

Ok, I'm convinced :+1: 
You mean something like:
```rust
impl App {
  fn set_advertisement_data(&self, ....) -> ReturnCode {
  // todo 
  }
  
  // more methods...
}
```

Yes, testability with less `void` functions is a good point but still the callbacks can't be testable!

Reply:
Yes, though since we're always going to enter the `App`, it _can_ be a mutable `self` reference if that's useful.